### PR TITLE
RavenDB-21106: Optimize the training for locality.

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-21106

### Additional description
IO deprived setups would incur high cost in dictionary pretraining. We are looking into optimizing it as much as possible with the objective of having to avoid configuration knobs to deal with it.

### Type of change
- Optimization

### How risky is the change?
- Low 

### Backward compatibility
- Non breaking change

### CODE REVIEW

File: src/Voron/Data/Tables/Table.cs [hunk 1]
```diff
@@ -1684,28 +1687,123 @@ public bool SeekOnePrimaryKeyWithPrefix(Slice prefix, Slice value, out TableValu
...
+                if (readDocuments == 0)
+                    return 0;
+
+                // We will sort the memory locations and try to maximize locality by sampling in chunks. 
+                chunkSpan = chunkSpan.Slice(0, readDocuments);
+                Sort.Run(chunkSpan);
+
+                // We will divide the whole set into chunks of size up to 128 skips. Less than that
+                // and the tradeoff will be bad for IO (which we are trying to limit as much as possible). 
+                int localChunkSize = Math.Max(128, ChunkSize / ((int)skip + 1));
```
Comment (ayende) [inline src/Voron/Data/Tables/Table.cs:1772]: What if you have less than 128?

File: src/Voron/Data/Tables/Table.cs [hunk 2]
```diff
@@ -1684,28 +1687,123 @@ public bool SeekOnePrimaryKeyWithPrefix(Slice prefix, Slice value, out TableValu
...
 
+            // Considering that in big collections the max amount of documents has been defined to be in the order of 100K 
+            // If we have to big skips we should get as many documents as possible to obtain as high locality as data allows.
+            int ChunkSize = 1024 * 16;
+
+            int ReadChunked(FixedSizeTree<long>.IFixedSizeIterator it, ByteString chunk)
+            {
+                // We will fill the buffer of the chunk expecting to have many documents in the chunk that end up 
+                // being stored in segments that are close to each other.
+                var chunkSpan = new Span<long>(chunk.Ptr, ChunkSize);
```
Comment (ayende) [inline src/Voron/Data/Tables/Table.cs:1746]: Add assert on the size here (in particular bytes vs number of elements)

File: src/Voron/Data/Tables/Table.cs [hunk 3]
```diff
@@ -1684,28 +1687,123 @@ public bool SeekOnePrimaryKeyWithPrefix(Slice prefix, Slice value, out TableValu
...
+                }
+
+                if (readDocuments == 0)
+                    return 0;
+
+                // We will sort the memory locations and try to maximize locality by sampling in chunks. 
+                chunkSpan = chunkSpan.Slice(0, readDocuments);
+                Sort.Run(chunkSpan);
+
+                // We will divide the whole set into chunks of size up to 128 skips. Less than that
```
Comment (ayende) [inline src/Voron/Data/Tables/Table.cs]: You read 16K entries, sort their locations and select 128?

Comment (redknightlois) [reply src/Voron/Data/Tables/Table.cs]: Min of 128.

File: src/Voron/Data/Tables/Table.cs (hunk 1 above)
Comment (redknightlois) [reply src/Voron/Data/Tables/Table.cs:1772]: We get 128. That is what it can hurt compression performance and what we need to monitor.

**Comment by redknightlois** [comment]
I made some mistake between what I tested and what I rebased and pushed. :(

**Comment by redknightlois** [comment]
@ayende Can you look into why `FastTests.Corax.DynamicFieldsIntegration.CannotChangeAnalyzerDuringIndexing` fails? There is something strange going on as the `it.CurrentKey` in the original implementation without the optimization it is always 2, but in here it is changing. Something is going on, but I need a second pair of eyes (I suspect some weird interplay of structs).

**Comment by redknightlois** [comment]
@arekpalinski Test this on the IO deprived setup. It seems to be improved so much that I don't know if we would need to add extra short-circuits like time.

**Comment by arekpalinski** [comment]
@redknightlois Please check my comment in the issue: https://issues.hibernatingrhinos.com/issue/RavenDB-21106/Corax-Dictionary-training-is-really-slow-on-initial-run-because-of-high-IO-usage#focus=Comments-67-722931.0-0

**Comment by redknightlois** [comment]
Current implementation will try to keep the amount of tokens homogeneous between indexes. This is particularly important for the case where we have big documents with full text search indexing. In those cases the amount of tokens may grow to the 100s of millions instead to be kept around 1M tokens. 

On local testing (Twitter small dataset) the index size stays roughly the same with the full token analysis showing that after certain amount of tokens the improvement achieved is negligible. While at the same time, the cost of training went down by at least half of the time.

File: src/Voron/Data/Tables/Table.cs [hunk 4]
```diff
@@ -1684,30 +1684,153 @@ public bool SeekOnePrimaryKeyWithPrefix(Slice prefix, Slice value, out TableValu
...
+            int ReadChunked(FixedSizeTree<long>.IFixedSizeIterator it, ByteString chunk, out bool hasMore)
             {
-                count++;
-                if (count % skip == 0)
+                // We will fill the buffer of the chunk expecting to have many documents in the chunk that end up 
+                // being stored in segments that are close to each other.
+                var chunkSpan = new Span<long>(chunk.Ptr, ChunkSize);
+                var keySpan = new Span<long>(chunk.Ptr + ChunkSize * sizeof(long), ChunkSize);
+                Debug.Assert(chunkSpan.Length == ChunkSize);
+                Debug.Assert(keySpan.Length == ChunkSize);
```
Comment (ayende) [inline src/Voron/Data/Tables/Table.cs:1750]: You are checking the length of the _spans_, but you need to check the size of _chunk_.

File: src/Voron/Data/Tables/Table.cs [hunk 5]
```diff
@@ -1684,30 +1684,153 @@ public bool SeekOnePrimaryKeyWithPrefix(Slice prefix, Slice value, out TableValu
...
+                if (readDocuments == 0)
+                    return 0;
+
+                Debug.Assert(readDocuments <= ChunkSize, "We cannot have more than ChunkSize because that is a buffer overflow.");
+
+                // We will divide the whole set into chunks of at least 128 skips (could be bigger if collection is small).
+                // Less than that and the tradeoff will be bad for IO (which we are trying to limit as much as possible). 
+                int localChunkSize = Math.Max(128, ChunkSize / ((int)skip + 1));
+                if (readDocuments < localChunkSize)
+                    return readDocuments;
```
Comment (ayende) [inline src/Voron/Data/Tables/Table.cs:1774]: This should not be sorted to?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Voron/Data/Tables/Table.cs (hunk 5 above)
Comment (arekpalinski) [reply src/Voron/Data/Tables/Table.cs:1774]: I believe this was already answered here:
https://github.com/ravendb/ravendb/commit/d4456d6975d06e3cba443ccfccec9a9c00d0ff46#r125409585

File: src/Voron/Data/Tables/Table.cs (hunk 4 above)
Comment (arekpalinski) [reply src/Voron/Data/Tables/Table.cs:1750]: @redknightlois ?

**Pull request merged**